src/query_plan/query_planner.rs

```diff
@@ -106,26 +120,303 @@ impl Default for QueryPlannerDebugConfig {
     /// A set of the names of interface or union types that have inconsistent "runtime types" across
     /// subgraphs.
     // PORT_NOTE: Named `inconsistentAbstractTypesRuntimes` in the JS codebase, which was slightly
     // confusing.
-    abstract_types_with_inconsistent_runtime_types: Arc<IndexSet<AbstractTypeDefinitionPosition>>,
+    abstract_types_with_inconsistent_runtime_types: IndexSet<AbstractTypeDefinitionPosition>,
     // TODO: Port _lastGeneratedPlanStatistics from the JS codebase in a way that keeps QueryPlanner
     // immutable.
 }
+
+impl QueryPlanner {
+    pub fn new(
+        supergraph: Supergraph,
+        config: QueryPlannerConfig,
+    ) -> Result<Self, FederationError> {
+        let supergraph_schema = supergraph.schema.clone();
+        let api_schema = supergraph.to_api_schema(ApiSchemaOptions {
+            include_defer: config.incremental_delivery.enable_defer,
+            ..Default::default()
+        })?;
+        let query_graph = build_federated_query_graph(
+            supergraph_schema.clone(),
+            api_schema.clone(),
+            Some(true),
+            Some(true),
+        )?;
+
+        let metadata = supergraph_schema.metadata().unwrap();
+
+        let federation_link = metadata.for_identity(&Identity::federation_identity());
+        let interface_object_directive =
+            federation_link.map_or(FEDERATION_INTERFACEOBJECT_DIRECTIVE_NAME_IN_SPEC, |link| {
+                link.directive_name_in_schema(&FEDERATION_INTERFACEOBJECT_DIRECTIVE_NAME_IN_SPEC)
+            });
+
+        let join_link = metadata.for_identity(&Identity::join_identity()).unwrap();
+        let join_field_directive =
+            join_link.directive_name_in_schema(&JOIN_FIELD_DIRECTIVE_NAME_IN_SPEC);
+
+        let is_interface_object =
+            |ty: &ExtendedType| ty.is_object() && ty.directives().has(&interface_object_directive);
+
+        let interface_types_with_interface_objects = supergraph
+            .schema
+            .get_types()
+            .filter_map(|position| match position {
+                TypeDefinitionPosition::Interface(interface_position) => Some(interface_position),
+                _ => None,
+            })
+            .filter(|position| {
+                query_graph.sources().any(|schema| {
+                    schema
+                        .schema()
+                        .types
+                        .get(&position.type_name)
+                        .is_some_and(is_interface_object)
+                })
+            })
+            .collect::<IndexSet<_>>();
+
+        let is_inconsistent = |position: AbstractTypeDefinitionPosition| {
+            let mut sources = query_graph.sources().filter_map(|subgraph| {
+                match subgraph.try_get_type(position.type_name().clone())? {
+                    TypeDefinitionPosition::Object(object) => Some(IndexSet::from([object])),
```

Comment:
The JS code skips object types (with a comment explaining why). Is there a reason to do something different here?